pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1395,3 +1390,87 @@ func updatePipelineRunStatusFromChildRefs(logger *zap.SugaredLogger, pr *v1beta1
+	// list VerificationPolicies for trusted resources
+	vp, err := c.verificationPolicyLister.VerificationPolicies(pr.Namespace).List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list VerificationPolicies from namespace %s with error %w", pr.Namespace, err)
+	}
+	var refSourceURI string
+	if resolvedTask.ResolvedTask.RefSource != nil {
+		refSourceURI = resolvedTask.ResolvedTask.RefSource.URI
+	}
+	// get matched policies
+	matchedPolicies, err := trustedresources.GetMatchedPolicies(ctx, refSourceURI, vp)
+	if err != nil {
+		pr.Status.MarkFailed(ReasonResourceVerificationFailed, err.Error())
+		return controller.NewPermanentError(err)
+	}
+	// verify if matched Policies and resolved task
+	if len(matchedPolicies) != 0 {
+		if resolvedTask.ResolvedTask.Task != nil {
+			err = trustedresources.VerifyTask(ctx, resolvedTask.ResolvedTask.Task, c.KubeClientSet, matchedPolicies)
+			if err != nil {
+				message := fmt.Sprintf("PipelineRun %s/%s referred task %s failed signature verification", pr.Namespace, pr.Name, taskName)
+				pr.Status.MarkFailed(ReasonResourceVerificationFailed, message)
+				return controller.NewPermanentError(err)
+			}
+		}
+	} else {
+		err := fmt.Errorf("PipelineRun %s/%s referred task %s doesn't have matched policies: %w", pr.Namespace, pr.Name, taskName, trustedresources.ErrNoMatchedPolicies)
+		switch config.GetVerificationNoMatchPolicy(ctx) {
+		case config.WarnNoMatchPolicy:
+			logger := logging.FromContext(ctx)
+			logger.Warnf(err.Error())
+		case config.FailNoMatchPolicy:
+			pr.Status.MarkFailed(ReasonResourceVerificationFailed, err.Error())
+			return controller.NewPermanentError(err)
+		}
+	}
+	return nil
+}
+
+func (c *Reconciler) verifyPipeline(ctx context.Context, pr *v1beta1.PipelineRun, pipelineMeta *resolutionutil.ResolvedObjectMeta, pipelineSpec *v1beta1.PipelineSpec) error {
+	// list VerificationPolicies for trusted resources
+	vp, err := c.verificationPolicyLister.VerificationPolicies(pr.Namespace).List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list VerificationPolicies from namespace %s with error %w", pr.Namespace, err)
+	}
+	var refSourceURI string
+	if pipelineMeta.RefSource != nil {
+		refSourceURI = pipelineMeta.RefSource.URI
+	}
+	// get matched policies
+	matchedPolicies, err := trustedresources.GetMatchedPolicies(ctx, refSourceURI, vp)
+	if err != nil {
+		pr.Status.MarkFailed(ReasonFailedValidation, err.Error())
+		return controller.NewPermanentError(err)
+	}
+	// verify if matched Policies
+	var pipelineToBeVerified *v1beta1.Pipeline
+	if pipelineMeta != nil {
+		pipelineToBeVerified = &v1beta1.Pipeline{
+			ObjectMeta: *pipelineMeta.ObjectMeta,
+			Spec:       *pipelineSpec,
+		}
+	}
+	if len(matchedPolicies) != 0 {
+		err = trustedresources.VerifyPipeline(ctx, pipelineToBeVerified, c.KubeClientSet, matchedPolicies)
```

Comment:
IIUC, the verification process breaks down to 4 steps:

1. Get policies
2. Fill ObjectMeta
3. Call Verify for type
4. Check error

1, 2, and 4 seems mostly copy and pasted across resolvedTask, Pipeline, and Task implementations. Can we refactor this further and pull this into a common implementation with the differing pieces configured as needed?

Reply:
The status setting is something that could also probably be pulled out - this way the verification logic doesn't need to be aware of the ins and outs of how statuses get set.

Another way to address this might be to move 1, 2, and 3 into the trustedresources package.

Reply:
Thanks! I refactored them into 1 function in trusted resources pkg

Reply:
I updated the pr again to only move the verification to `reconcile` reduce the code changes